Add the kvm-libvirt platform to lokoctl

pothos opened the pull request:
The KVM libvirt Terraform module was not yet available from lokoctl
despite it being an easy way to try out Lokomotive without any cloud
provider accounts (and their attached costs). It allows to use
Lokomotive on any Flatcar Container Linux image, even local development
builds. It also gives direct access to the VGA console for debugging.
The cluster VMs can be shut down in virt-manager while they are not
needed.

(Reactivating the CI pipeline for libvirt is the next step.)

Closes: https://github.com/kinvolk/lokomotive/issues/214

pothos committed:
Add the kvm-libvirt platform to lokoctl

The KVM libvirt Terraform module was not yet available from lokoctl
despite it being an easy way to try out Lokomotive without any cloud
provider accounts (and their attached costs). It allows to use
Lokomotive on any Flatcar Container Linux image, even local development
builds. It also gives direct access to the VGA console for debugging.
The cluster VMs can be shut down in virt-manager while they are not
needed.

johananl commented:
Thanks @pothos! I'm looking at this. In the meantime, could you please fix the linting issues? If you don't have time to follow up, we can pick this up, too.

pothos commented:
Thanks for having a look. I can address some of the issues but I'm not sure about all of the linting issues because similar code is present in other Lokomotive platforms already (without disabling the linter for these lines).

johananl commented:
> Thanks for having a look. I can address some of the issues but I'm not sure about all of the linting issues because similar code is present in other Lokomotive platforms already (without disabling the linter for these lines).

Yeah, the linter only checks added/modified lines. I suggest we fix the linting issues here and I can open a follow up issue to do the same for the other platforms. We need a green CI before we can merge.

johananl reviewed (changes requested):
Thanks a lot for the PR. On the technical side this seems to work very nicely.

I have multiple issues with the docs - see my inline comments.

pothos commented on ci/kvm-libvirt/kvm-libvirt-cluster.lokocfg.envsubst:
```diff
@@ -0,0 +1,11 @@
+cluster "kvm-libvirt" {
```
There was one which shared the baremetal CI setup. The code was probably removed but can be recovered from git.

pothos commented:
Thanks, will try to go through it end of the week.

johananl commented:
> Thanks, will try to go through it end of the week.

Cool. As said, we can also help with matching the style to our existing guides.

pothos commented:
Feel free to push commits :+1:

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+If you need a reproducible setup, please add ?ref=HASH to the end of
+each `git::` path in the `source = "git::…"` imports.
+
+You can consider to change the Terraform modules provided by Lokomotive
```
Oh, this is outdated a outdated copy-paste and I forgot to review this, thanks for catching

pothos commented on assets/lokomotive-kubernetes/kvm-libvirt/flatcar-linux/kubernetes/outputs.tf:
```diff
@@ -26,6 +26,14 @@ output "libvirtbaseid" {
 }
 
+output "cluster_domain_suffix" {
+  value = var.cluster_domain_suffix
```
The worker modules use them to have only one place where the value is fed in and not two. Since this is now behind lokoctl we can change this.

pothos replied:
done

pothos commented on docs/quickstarts/kvm-libvirt.md:
````diff
@@ -0,0 +1,272 @@
+sure that a couple of container images fit on the node.
+
+```sh
+$ sudo dnf install wget bzip2 qemu-img  # or sudo apt install wget bzip2 qemu-utils
````
Ok, should this be moved to the bullet point where the requirements are stated, like _Fedora packages A B C or Debian/Ubuntu packages A B C_?

johananl replied:
IMO it could make sense to list per-distro packages for common distros in the requirements if that's the question.

pothos replied:
I've updated the PR and sticked to the steps with the distro comments.

pothos commented on pkg/platform/kvmlibvirt/kvmlibvirt.go:
```diff
@@ -0,0 +1,198 @@
+	ClusterName                  string       `hcl:"cluster_name"`
+	ControllerCount              int          `hcl:"controller_count,optional"`
+	MachineDomain                string       `hcl:"machine_domain"`
+	OSImageUnpacked              string       `hcl:"os_image_unpacked"`
```
Thanks, will change it

pothos commented on docs/quickstarts/kvm-libvirt.md:
````diff
@@ -0,0 +1,272 @@
+sure that a couple of container images fit on the node.
+
+```sh
+$ sudo dnf install wget bzip2 qemu-img  # or sudo apt install wget bzip2 qemu-utils
````
Ok, quite confusing

johananl replied:
You mean that not having dollar signs is confusing?

IMO not necessarily. An example without dollar signs that is very clear IMO: https://github.com/kinvolk/lokomotive/blob/master/docs/quickstarts/packet.md#verification

The reason I personally don't like dollar signs is that they interfere with copy-paste.

pothos replied:
Ok, I didn't separate the output if it's just one command or somehow clear what is what.

pothos commented on docs/configuration-reference/platforms/kvm-libvirt.md:
```diff
@@ -0,0 +1,142 @@
+| `enable_aggregation`                  | Enable the Kubernetes Aggregation Layer.                                                                                                                                                                                                                                          |        true        |     bool     |  false   |
+| `enable_reporting`                    | Enables usage or analytics reporting to upstream.                                                                                                                                                                                                                                 |        false       |     bool     |  false   |
+| `kube_apiserver_extra_flags`          | Extra flags to pass to the kube-apiserver binary.                                                                                                                                                                                                                                 |          []        | list(string) |  false   |
+| `machine_domain`                      | DNS zone.                                                                                                                                                                                                                                                                         |          -         |    string    |   true   |
```
Thanks, updated

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+  cluster_name = "vmcluster"
+  machine_domain = "vmcluster.k8s"
+  # Path to where the image was prepared, note the triple slash for the absolute path
+  os_image_unpacked = "file:///PATH/TO/flatcar_production_qemu_image.img"
```
done

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+  ssh_pubkeys = [file(pathexpand("~/.ssh/id_rsa.pub"))]
+  cluster_name = "vmcluster"
+  machine_domain = "vmcluster.k8s"
+  # Path to where the image was prepared, note the triple slash for the absolute path
```
done

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+
+## Introduction
+
+In this guide, we'll create a Lokomotive cluster with Flatcar Container Linux VMs on our local machine.
```
done

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+## Introduction
+
+In this guide, we'll create a Lokomotive cluster with Flatcar Container Linux VMs on our local machine.
+The cluster will consist of controller and worker nodes.
```
thanks, should be gone now

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+
+Controllers are provisioned to run an `etcd-member` peer and a `kubelet` service.
+Workers run just a `kubelet` service. A one-time [bootkube](https://github.com/kubernetes-incubator/bootkube)
+bootstrap schedules the `apiserver`, `scheduler`, `controller-manager`, and `coredns` on controllers
```
thanks, taken

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+## Requirements
+
+* A Linux host OS with `qemu-img`, `wget`, and `bunzip2` binaries installed
+* At least 4 GB of (free) RAM
```
thanks

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+
+## Requirements
+
+* A Linux host OS with `qemu-img`, `wget`, and `bunzip2` binaries installed
```
I added the periods

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+* A Linux host OS with `qemu-img`, `wget`, and `bunzip2` binaries installed
+* At least 4 GB of (free) RAM
+* An SSH public and private key
+* Running libvirtd system service, see [libvirtd Setup](#step-3-libvirtd-setup)
```
removed it from requirements

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+
+Ensure that libvirtd is set up.
+The examples in this guide use Fedora's `dnf`
+but for example in Ubuntu you would use `apt` as stated in the comment in the same line.
```
done

pothos commented on pkg/platform/kvmlibvirt/kvmlibvirt.go:
```diff
@@ -0,0 +1,198 @@
+type workerPool struct {
+	Name          string   `hcl:"pool_name,label"`
+	Count         int      `hcl:"count"`
+	VirtualCpus   int      `hcl:"virtual_cpus,optional"`
```
done

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+
+Do not use this image file with QEMU directly because it needs to stay clean for Ignition to run on first boot.
+
+### Step 3: libvirtd Setup
```
I think it's a nicer experience if we cover the steps and don't leave it to the user to know if their system fulfills the requirements.

pothos commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,272 @@
+see the [configuration reference](../configuration-reference/platforms/kvm-libvirt.md).
+
+
+### Step 6: Check ssh-agent
```
I shortened it a bit but it doesn't really fit into the next section.

pothos commented:
Rebased the PR but now bootstrap reports this error and times out: `WARNING: /assets/manifests does not exist, not creating any self-hosted assets`

ipochi commented:
> Rebased the PR but now bootstrap reports this error and times out: `WARNING: /assets/manifests does not exist, not creating any self-hosted assets`

Thats because no charts are copied for installion.

I have got it working, plus

* rebased to the latest master
* tls bootstrapping enabled

Will update the PR tomorrow after cleanup.

Commit:
platform/kvmlibvirt: add charts to be installed

This commit adds the control plane charts that need to be installed once
the bootstrap control plane is running.

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
terraform: update terraform providers to new format

This commit updates the versions.tf and terraform providers to the
required format for 0.13 version.

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
kvmlibvirt: enable tls bootstrap

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
lint: fix lint issues

Signed-off-by: Imran Pochi <imran@kinvolk.io>

ipochi commented:
@invidian @johananl @surajssd 
PR is open for re-review.

Thanks to @pothos for initiating the effort around this.

knrt10 commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,284 @@
+* A Linux host OS.
+* At least 4 GB of available RAM.
+* An SSH key pair.
+* Terraform `v0.12.x`
```
We now need `v0.13.x`

knrt10 commented on docs/quickstarts/kvm-libvirt.md:
```diff
@@ -0,0 +1,284 @@
+* An SSH key pair.
+* Terraform `v0.12.x`
+  [installed](https://learn.hashicorp.com/terraform/getting-started/install.html#install-terraform).
+* [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct),
```
No need for this, it gets automatically installed now

invidian replied:
#1078 :smile:

knrt10 commented on pkg/platform/kvmlibvirt/kvmlibvirt.go:
```diff
@@ -0,0 +1,198 @@
+	VirtualCPUs   int      `hcl:"virtual_cpus,optional"`
+	VirtualMemory int      `hcl:"virtual_memory,optional"`
+	CLCSnippets   []string `hcl:"clc_snippets,optional"`
+	Labels        string   `hcl:"labels,optional"`
```
We should use `map[string]string` as we are doing with other platforms

knrt10 commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,186 @@
+  virtual_memory {{ $pool.VirtualMemory }}
+  {{- end }}
+
+  {{- if $pool.Labels }}
+  labels = {{ $pool.Labels }}
+  {{- end }}
```
```
{{- if $pool.Labels }}
  labels = {
  {{- range $k, $v := $pool.Labels }}
    "{{ $k }}" = "{{ $v }}",
  {{- end }}
 }
```

knrt10 commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,186 @@
+
+provider "random" {
+  version = "~> 2.2"
+}
```
I guess this can be removed from here and moved to `assets/terraform-modules/kvm-libvirt/flatcar-linux/kubernetes/versions.tf`

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+provider "libvirt" {
+	uri     = "qemu:///system"
```
Just started setting up Terraform with existing libvirtd cluster (not lokomotive) and noticed this PR.

For those of us with a home 'server' it would be great if we can pass the kvm host through cluster configuration, perhaps along the lines of:
```
  {{- if .Config.QemuURI }}
  uri     = {{.Config.QemuURI }}
  {{else}} 
  uri     = "qemu:///system"
  {{- end}}
```

pothos replied:
To use a remote location? Maybe it's not easy to get this works with the image volume pool, or?

kskewes replied:
Providing a:
1. external file eg: `https://example.com/flat-car.img` results in terraform runner (laptop) downloading it and then copying across the network to the server. I think this is reasonable.
2. laptop local file eg: `/path/to/flat-car.img` results in terraform runner copying across network to the server. Also reasonable.

It is unintuitive but I don't think there's an alternative - see: https://github.com/dmacvicar/terraform-provider-libvirt/issues/299#issuecomment-419676846

I currently supply a path to images that's actually a NFS mount from the 'server' which is kind of gross (looping data) but works in the sense that I don't need to keep the image on my laptop. :sweat_smile: 

```
# Qcow2 base OS volume that can be shared by VM's (domain's)
resource "libvirt_volume" "os_base" {
  name   = "${var.guest_hostname}-os_base"
  source = "/mnt/nfs/path/to/ubuntu-focal-server-cloudimg-amd64.img"
}

# OS volume per guest VM
resource "libvirt_volume" "os" {
  count          = var.guest_count
  name           = "${var.guest_hostname}-${format("%02d", count.index)}-os.qcow2"
  base_volume_id = libvirt_volume.os_base.id
  pool           = var.guest_pool_name
  size           = var.os_volume_size
}
```

kskewes replied:
Please let me know if I misunderstood the question.
If the concept is amenable, perhaps we could add a line to markdown file saying that image sources will be fetched by lokoctl executor and copied to remote kvm host?

Maybe I build a fork of this PR with the change and do some testing first. :)
Then I can do a small PR once this one merged or just post here that it works as suspected if this is still being iterated on?

kskewes replied:
I started building this PR and trying to run it against my remote kvm server.

On top of a couple of string quotes suggested here the changes are: https://github.com/kskewes/lokomotive/commit/6246cc76e8e7287759023565e5fc000677aa4266
It creates all of the libvirt domains and things but the default NAT networking doesn't enable access from lokoctl runner to hosts. It really needs to be a bridge network for me. I'll have another go later and can iterate in a fork.

kskewes commented on pkg/platform/kvmlibvirt/kvmlibvirt.go:
```diff
@@ -0,0 +1,235 @@
+	DisableSelfHostedKubelet     bool         `hcl:"disable_self_hosted_kubelet,optional"`
+	EnableReporting              bool         `hcl:"enable_reporting,optional"`
+	EnableAggregation            bool         `hcl:"enable_aggregation,optional"`
+	EnableTLSBootstrap           bool         `hcl:"enable_tsl_bootstrap,optional"`
```
```suggestion
	EnableTLSBootstrap           bool         `hcl:"enable_tls_bootstrap,optional"`
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  {{- end }}
+
+  {{- if $pool.VirtualMemory }}
+  virtual_memory {{ $pool.VirtualMemory }}
```
```suggestion
  virtual_memory = {{ $pool.VirtualMemory }}
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  {{- end }}
+
+  {{- if .Config.ControllerVirtualMemory}}
+  virtual_memory {{ .Config.ControllerVirtualMemory}}
```
```suggestion
  virtual_memory = {{ .Config.ControllerVirtualMemory}}
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  {{- end }}
+
+  {{- if .Config.PodCidr }}
+  pod_cidr = {{.Config.PodCidr }}
```
```suggestion
  pod_cidr = "{{.Config.PodCidr }}"
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  pod_cidr = {{.Config.PodCidr }}
+  {{- end }}
+  {{- if .Config.ServiceCidr }}
+  service_cidr = {{.Config.ServiceCidr }}
```
```suggestion
  service_cidr = "{{.Config.ServiceCidr }}"
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  cluster_domain_suffix = {{$.Config.ClusterDomainSuffix }}
+  {{- end }}
+  {{- if $.Config.ServiceCidr }}
+  service_cidr = {{$.Config.ServiceCidr }}
```
```suggestion
  service_cidr = "{{$.Config.ServiceCidr }}"
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  {{- end }}
+
+  {{- if .Config.ClusterDomainSuffix }}
+  cluster_domain_suffix = {{.Config.ClusterDomainSuffix }}
```
```suggestion
  cluster_domain_suffix = "{{.Config.ClusterDomainSuffix }}"
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  machine_domain        = "{{$.Config.MachineDomain}}"
+  cluster_name          = "{{$.Config.ClusterName}}"
+  {{- if $.Config.ClusterDomainSuffix }}
+  cluster_domain_suffix = {{$.Config.ClusterDomainSuffix }}
```
```suggestion
  cluster_domain_suffix = "{{$.Config.ClusterDomainSuffix }}"
```

kskewes commented on pkg/platform/kvmlibvirt/template.go:
```diff
@@ -0,0 +1,190 @@
+  {{- end }}
+
+  {{- if .Config.NetworkIPAutodetectionMethod }}
+  network_ip_autodetection_method = {{ .Config.NetworkIPAutodetectionMethod }}
```
```suggestion
  network_ip_autodetection_method = "{{ .Config.NetworkIPAutodetectionMethod }}"
```

invidian commented:
@kskewes if you're interested in Lokomotive using libvirt underneath, there is also #392, which might be in a better shape than this PR (though libvirt usage there is rather only for testing).

kskewes commented:
> @kskewes if you're interested in Lokomotive using libvirt underneath, there is also #392, which might be in a better shape than this PR (though libvirt usage there is rather only for testing).

Had a go at running up #392.
1. I like the de-duplication of terraform modules.
2. It seems to rely on known network addressing for `tinkerbell` which is possible with NAT networks, matching static mac allocation on my router, or plumbing through some more variables.
3. NAT networks in libvirt allow host originated connections to ssh to VM's, but it doesn't (by default) setup iptables rules for remote hosts (my laptop/`lokoctl` runner). Currently I bridge separate NIC's on the kvm host switch ports and this keeps the kvm host networking relatively clean. 

Ultimately I think either continuing with #810 or doing my own libvirt plus [bare-metal quickstart](https://github.com/kinvolk/lokomotive/blob/master/docs/quickstarts/baremetal.md) might be the fastest way for me to come up to speed on flatcar linux and lokomotive. Then look at how I might be able to assist with #392.
